File: src/sources/git.rs
use super::{ProgressReporter, Source};
use std::io::Read;
use std::path::{Path, PathBuf};
use std::process::Command;

/// Clone or pull a git repository into the staging directory.
pub struct GitSource {
    pub url: String,
    pub auth: Option<String>,
    pub ref_name: String,
}

#[async_trait::async_trait]
impl Source for GitSource {
    async fn sync_to_local(
        &self,
        staging_root: &Path,
        progress: &mut dyn ProgressReporter,
    ) -> Result<PathBuf, Box<dyn std::error::Error + Send + Sync>> {
        let dir_name = super::uri_staging_dir(&super::SourceUri::Git {
            url: self.url.clone(),
        });
        let local_dir = staging_root.join(&dir_name);

        let git_dir = local_dir.join(".git");
        if git_dir.exists() {
            progress.report(&format!(
                "git repo exists at {}, pulling {}...",
                local_dir.display(),
                self.ref_name
            ));
            fetch_and_checkout(&local_dir, &self.ref_name, progress)?;
        } else {
            tokio::fs::create_dir_all(&local_dir).await?;
            progress.report(&format!("cloning {} (ref: {})...", self.url, self.ref_name));
            let clone_url = maybe_inject_auth(&self.url, self.auth.as_deref());
            clone_repo(&clone_url, &local_dir, &self.ref_name, progress)?;
        }

        Ok(local_dir)
    }

    fn name(&self) -> &str {
        "git"
    }

    /// Poll remote tip SHA via `git ls-remote` without cloning.
    async fn remote_fingerprint(
        &self,
    ) -> Result<Option<String>, Box<dyn std::error::Error + Send + Sync>> {
        let fetch_url = maybe_inject_auth(&self.url, self.auth.as_deref());
        let output = tokio::process::Command::new("git")
            .args(["ls-remote", &fetch_url, &self.ref_name])
            .output()
            .await
            .map_err(|e| format!("git ls-remote failed: {}", e))?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            return Err(format!("git ls-remote failed: {}", stderr).into());
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        // First field of first line is the commit SHA
        let sha = stdout.split_whitespace().next().unwrap_or("").to_string();
        if sha.is_empty() {
            return Ok(None);
        }
        Ok(Some(sha))
    }

    /// Materialize content without persistent clone: try `git archive --remote`
    /// first, fall back to shallow clone.
    async fn materialize_ephemeral(
        &self,
        staging_root: &Path,
        progress: &mut dyn ProgressReporter,
    ) -> Result<PathBuf, Box<dyn std::error::Error + Send + Sync>> {
        let dir_name = super::uri_staging_dir(&super::SourceUri::Git {
            url: self.url.clone(),
        });
        let local_dir = staging_root.join(&dir_name);

        let fetch_url = maybe_inject_auth(&self.url, self.auth.as_deref());

        // Try git archive --remote first (no clone, pure HTTP archive)
        progress.report(&format!(
            "downloading archive for {} @ {}...",
            self.url, self.ref_name
        ));
        if let Ok(()) = archive_extract(&fetch_url, &self.ref_name, &local_dir, progress) {
            return Ok(local_dir);
        }

        // Fall back to shallow clone (still avoids full history)
        progress.report("archive download failed; falling back to shallow clone");
        if local_dir.exists() {
            std::fs::remove_dir_all(&local_dir).ok();
        }
        tokio::fs::create_dir_all(&local_dir).await?;
        clone_repo(&fetch_url, &local_dir, &self.ref_name, progress)?;
        Ok(local_dir)
    }
}

fn maybe_inject_auth(url: &str, auth: Option<&str>) -> String {
    let Some(token) = auth else {
        return url.to_string();
    };
    // For https URLs, inject the token as userinfo.
    if let Some(rest) = url.strip_prefix("https://") {
        return format!("https://oauth2:{}@{}", token, rest);
    }
    url.to_string()
}

fn clone_repo(
    url: &str,
    dir: &Path,
    ref_name: &str,
    progress: &mut dyn ProgressReporter,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    progress.report(&format!("git clone --depth 1 --branch {} ...", ref_name));

    let mut cmd = Command::new("git");
    cmd.args(["clone", "--depth", "1", "--branch", ref_name, url])
        .arg(dir);

    let output = cmd
        .output()
        .map_err(|e| format!("git clone failed: {}", e))?;

    if !output.status.success() {
        let _stderr = String::from_utf8_lossy(&output.stderr);
        // Fallback: clone default branch then checkout.
        progress.report("shallow clone failed, trying full clone + checkout...");
        let mut cmd2 = Command::new("git");
        cmd2.args(["clone", url]).arg(dir);
        let out2 = cmd2
            .output()
            .map_err(|e| format!("git clone fallback failed: {}", e))?;
        if !out2.status.success() {
            return Err(format!(
                "git clone fallback failed: {}",
                String::from_utf8_lossy(&out2.stderr)
            )
            .into());
        }
    }

    // Ensure we're on the right ref.
    fetch_and_checkout(dir, ref_name, progress)?;
    Ok(())
}

fn fetch_and_checkout(
    repo_dir: &Path,
    ref_name: &str,
    progress: &mut dyn ProgressReporter,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    // Fetch all to pick up new branches/tags.
    let fetch = Command::new("git")
        .current_dir(repo_dir)
        .args(["fetch", "--all", "--prune"])
        .output()
        .map_err(|e| format!("git fetch failed: {}", e))?;

    if !fetch.status.success() {
        return Err(format!(
            "git fetch failed: {}",
            String::from_utf8_lossy(&fetch.stderr)
        )
        .into());
    }

    progress.report(&format!("git checkout {} ...", ref_name));
    let checkout = Command::new("git")
        .current_dir(repo_dir)
        .args(["checkout", ref_name])
        .output()
        .map_err(|e| format!("git checkout failed: {}", e))?;

    if !checkout.status.success() {
        // Try as a remote branch reference.
        let remote_ref = format!("origin/{}", ref_name);
        progress.report(&format!("trying remote ref {}...", remote_ref));
        let co2 = Command::new("git")
            .current_dir(repo_dir)
            .args(["checkout", "-b", ref_name, &remote_ref])
            .output()
            .map_err(|e| format!("git checkout remote ref failed: {}", e))?;

        if !co2.status.success() {
            return Err(format!(
                "git checkout {} failed: {} (tried {}: {})",
                ref_name,
                String::from_utf8_lossy(&checkout.stderr),
                remote_ref,
                String::from_utf8_lossy(&co2.stderr)
            )
            .into());
        }
    }

    // Pull latest if on a branch (not detached HEAD).
    let pull = Command::new("git")
        .current_dir(repo_dir)
        .args(["pull", "--ff-only"])
        .output();

    match pull {
        Ok(o) if o.status.success() => progress.report("pulled latest"),
        Ok(o) => {
            let stderr = String::from_utf8_lossy(&o.stderr);
            progress.report(&format!("git pull skipped ({}): {}", o.status, stderr));
            // Shallow depth may prevent ff-only pull. Attempt to reset to
            // the remote tracking branch to pick up new commits.
            progress.report("attempting fetch --unshallow + reset...");
            let _ = Command::new("git")
                .current_dir(repo_dir)
                .args(["fetch", "--unshallow"])
                .output();
            let rebase = Command::new("git")
                .current_dir(repo_dir)
                .args(["merge", "--ff-only", &format!("origin/{}", ref_name)])
                .output();
            match rebase {
                Ok(r) if r.status.success() => progress.report("fast-forward after unshallow"),
                _ => {
                    progress.report("reset to origin/HEAD as fallback");
                    let reset = Command::new("git")
                        .current_dir(repo_dir)
                        .args(["reset", "--hard", &format!("origin/{}", ref_name)])
                        .output();
                    if let Ok(r) = reset {
                        if r.status.success() {
                            progress.report("reset to origin/HEAD OK");
                        }
                    }
                }
            }
        }
        Err(e) => progress.report(&format!("git pull error (non-fatal): {}", e)),
    }

    Ok(())
}

/// Download a git archive (tar) for the given ref and extract it into `dest`.
/// Uses `git archive --remote` which is a no-clone operation over HTTP/SSH.
/// Returns Err if the remote does not support git archive or the ref is invalid.
fn archive_extract(
    url: &str,
    ref_name: &str,
    dest: &Path,
    progress: &mut dyn ProgressReporter,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    if dest.exists() {
        std::fs::remove_dir_all(dest).map_err(|e| format!("remove old staging: {}", e))?;
    }
    std::fs::create_dir_all(dest).map_err(|e| format!("create staging dir: {}", e))?;

    let mut child = std::process::Command::new("git")
        .args(["archive", "--remote", url, ref_name, "--format=tar"])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| format!("git archive --remote spawn failed: {}", e))?;

    let tar_output = std::process::Command::new("tar")
        .args(["-xC", &dest.to_string_lossy()])
        .stdin(std::process::Stdio::from(child.stdout.take().unwrap()))
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()
        .map_err(|e| format!("tar extraction failed: {}", e))?;

    let archive_status = child
        .wait()
        .map_err(|e| format!("git archive wait: {}", e))?;

    if !archive_status.success() || !tar_output.status.success() {
        std::fs::remove_dir_all(dest).ok();
        let err_msg = if !archive_status.success() {
            // Read stderr from the child process that has already completed
            let archive_stderr = child
                .stderr
                .take()
                .and_then(|mut s| {
                    let mut buf = String::new();
                    s.read_to_string(&mut buf).ok().map(|_| buf)
                })
                .unwrap_or_default();
            let archive_stderr = if archive_stderr.is_empty() {
                "git archive failed with unknown error"
            } else {
                &archive_stderr
            };
            format!("git archive failed: {}", archive_stderr)
        } else {
            let stderr = String::from_utf8_lossy(&tar_output.stderr);
            format!("tar extraction failed: {}", stderr)
        };
        return Err(err_msg.into());
    }

    progress.report(&format!("archive extracted to {}", dest.display()));
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_maybe_inject_auth_no_token() {
        assert_eq!(
            maybe_inject_auth("https://github.com/user/repo.git", None),
            "https://github.com/user/repo.git"
        );
    }

    #[test]
    fn test_maybe_inject_auth_with_token() {
        assert_eq!(
            maybe_inject_auth("https://github.com/user/repo.git", Some("ghp_token123")),
            "https://oauth2:ghp_token123@github.com/user/repo.git"
        );
    }

    #[test]
    fn test_maybe_inject_auth_ssh_unchanged() {
        assert_eq!(
            maybe_inject_auth("git@github.com:user/repo.git", Some("token")),
            "git@github.com:user/repo.git"
        );
    }
}

File: go/internal/sources/git.go
// Git source (Rust src/sources/git.rs): clone/fetch a repository into the
// staging directory, poll the remote tip without cloning, and materialize a
// tree without a persistent clone.
//
// Every git invocation is argv-only (no shell) and bounded: `--depth 1` keeps
// the clone shallow, the ref is pinned, and each call gets a deadline from
// LEANKG_GIT_TIMEOUT seconds (default 10 minutes) so an unreachable host fails
// instead of hanging the index/watch loop. The Rust reference had no timeout.
package sources

import (
	"archive/tar"
	"bytes"
	"context"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// defaultGitTimeout bounds one git invocation when LEANKG_GIT_TIMEOUT is unset.
const defaultGitTimeout = 10 * time.Minute

// GitSource is a `git+<transport url>` source.
type GitSource struct {
	URL     string
	Auth    string
	RefName string
}

// Name implements Source.
func (g *GitSource) Name() string { return "git" }

// SyncToLocal clones the repository on first use and fetches/checkouts the ref
// afterwards (Rust git.rs sync_to_local).
func (g *GitSource) SyncToLocal(ctx context.Context, stagingRoot string, progress ProgressReporter) (string, error) {
	localDir, err := stagingPath(stagingRoot, g.uri())
	if err != nil {
		return "", err
	}
	if info, statErr := os.Stat(filepath.Join(localDir, ".git")); statErr == nil && info.IsDir() {
		report(progress, fmt.Sprintf("git repo exists at %s, pulling %s...", localDir, g.RefName))
		if err := fetchAndCheckout(ctx, localDir, g.RefName, progress); err != nil {
			return "", err
		}
		return localDir, nil
	}
	if err := os.MkdirAll(localDir, 0o755); err != nil {
		return "", err
	}
	report(progress, fmt.Sprintf("cloning %s (ref: %s)...", g.URL, g.RefName))
	if err := cloneRepo(ctx, injectAuth(g.URL, g.Auth), localDir, g.RefName, progress); err != nil {
		return "", err
	}
	return localDir, nil
}

// RemoteFingerprint polls the remote tip SHA with `git ls-remote` (no clone).
// An empty result means the ref could not be resolved (a bare SHA ref, or a
// remote that does not advertise it) and yields "" — Rust returned Ok(None).
func (g *GitSource) RemoteFingerprint(ctx context.Context) (string, error) {
	stdout, stderr, err := runGit(ctx, "", "ls-remote", injectAuth(g.URL, g.Auth), g.RefName)
	if err != nil {
		return "", gitError("git ls-remote failed", stderr, err)
	}
	fields := strings.Fields(stdout)
	if len(fields) == 0 {
		return "", nil
	}
	return fields[0], nil
}

// MaterializeEphemeral tries `git archive --remote` (no clone) and falls back
// to a shallow clone, matching Rust git.rs materialize_ephemeral.
func (g *GitSource) MaterializeEphemeral(ctx context.Context, stagingRoot string, progress ProgressReporter) (string, error) {
	localDir, err := stagingPath(stagingRoot, g.uri())
	if err != nil {
		return "", err
	}
	fetchURL := injectAuth(g.URL, g.Auth)

	report(progress, fmt.Sprintf("downloading archive for %s @ %s...", g.URL, g.RefName))
	if err := archiveExtract(ctx, fetchURL, g.RefName, localDir, progress); err == nil {
		return localDir, nil
	}

	report(progress, "archive download failed; falling back to shallow clone")
	if err := os.RemoveAll(localDir); err != nil {
		return "", err
	}
	if err := os.MkdirAll(localDir, 0o755); err != nil {
		return "", err
	}
	if err := cloneRepo(ctx, fetchURL, localDir, g.RefName, progress); err != nil {
		return "", err
	}
	return localDir, nil
}

func (g *GitSource) uri() URI { return URI{Kind: KindGit, URL: g.URL} }

// injectAuth injects a token into an https URL as `oauth2:<token>@host`
// (Rust maybe_inject_auth). Empty auth, and every non-https URL, pass through
// unchanged.
func injectAuth(url, auth string) string {
	if auth == "" {
		return url
	}
	if rest, ok := strings.CutPrefix(url, "https://"); ok {
		return "https://oauth2:" + auth + "@" + rest
	}
	return url
}

// cloneRepo does a shallow, branch-pinned clone and falls back to a full clone
// plus checkout when the ref is not a branch tip (a tag or a bare SHA).
func cloneRepo(ctx context.Context, url, dir, refName string, progress ProgressReporter) error {
	report(progress, fmt.Sprintf("git clone --depth 1 --branch %s ...", refName))
	if _, stderr, err := runGit(ctx, "", "clone", "--depth", "1", "--branch", refName, url, dir); err != nil {
		// Fallback: clone the default branch, then check the ref out.
		// Rust reported only the fallback's stderr; keep the shallow failure
		// in the progress stream so a real cause is still visible.
		report(progress, "shallow clone failed ("+firstLine(stderr, err)+"), trying full clone + checkout...")
		if _, stderr2, err2 := runGit(ctx, "", "clone", url, dir); err2 != nil {
			return gitError("git clone fallback failed", stderr2, err2)
		}
	}
	return fetchAndCheckout(ctx, dir, refName, progress)
}

// fetchAndCheckout fetches every remote ref, checks the requested ref out
// (branch, then origin/<ref>), and fast-forwards an existing branch. The
// pull/merge/reset tail is non-fatal in Rust and stays non-fatal here.
func fetchAndCheckout(ctx context.Context, repoDir, refName string, progress ProgressReporter) error {
	if _, stderr, err := runGit(ctx, repoDir, "fetch", "--all", "--prune"); err != nil {
		return gitError("git fetch failed", stderr, err)
	}

	report(progress, "git checkout "+refName+" ...")
	if _, checkoutStderr, err := runGit(ctx, repoDir, "checkout", refName); err != nil {
		remoteRef := "origin/" + refName
		report(progress, "trying remote ref "+remoteRef+" ...")
		if _, remoteStderr, err2 := runGit(ctx, repoDir, "checkout", "-b", refName, remoteRef); err2 != nil {
			return fmt.Errorf("git checkout %s failed: %s (tried %s: %s)",
				refName, strings.TrimSpace(checkoutStderr), remoteRef, strings.TrimSpace(remoteStderr))
		}
	}

	// A shallow depth can block a fast-forward pull: unshallow, then
	// fast-forward, then hard-reset to the remote tracking branch.
	_, pullStderr, pullErr := runGit(ctx, repoDir, "pull", "--ff-only")
	if pullErr == nil {
		report(progress, "pulled latest")
		return nil
	}
	report(progress, "git pull skipped ("+firstLine(pullStderr, pullErr)+"), attempting fetch --unshallow + reset...")
	if _, _, err := runGit(ctx, repoDir, "fetch", "--unshallow"); err != nil {
		report(progress, "fetch --unshallow failed; keeping the existing worktree")
		return nil
	}
	if _, _, err := runGit(ctx, repoDir, "merge", "--ff-only", "origin/"+refName); err == nil {
		report(progress, "fast-forward after unshallow")
		return nil
	}
	report(progress, "reset to origin/HEAD as fallback")
	if _, _, err := runGit(ctx, repoDir, "reset", "--hard", "origin/"+refName); err == nil {
		report(progress, "reset to origin/HEAD OK")
	}
	return nil
}

// archiveExtract streams `git archive --remote <url> <ref> --format=tar` into
// dest. It fails when the remote does not support the archive protocol (the
// usual case for GitHub over https), which makes the caller fall back to a
// shallow clone. Rust piped into the system tar; the archive/tar stdlib reader
// removes the external-tar dependency and lets entry paths be validated.
func archiveExtract(ctx context.Context, url, refName, dest string, progress ProgressReporter) error {
	if err := os.RemoveAll(dest); err != nil {
		return fmt.Errorf("remove old staging: %w", err)
	}
	if err := os.MkdirAll(dest, 0o755); err != nil {
		return fmt.Errorf("create staging dir: %w", err)
	}

	archCtx, cancel := gitContext(ctx)
	defer cancel()
	cmd := exec.CommandContext(archCtx, "git", "archive", "--remote", url, refName, "--format=tar")
	stderr := &bytes.Buffer{}
	cmd.Stderr = stderr
	pipe, err := cmd.StdoutPipe()
	if err != nil {
		return gitError("git archive --remote failed", "", err)
	}
	if err := cmd.Start(); err != nil {
		return gitError("git archive --remote spawn failed", stderr.String(), err)
	}
	extractErr := extractTar(pipe, dest)
	waitErr := cmd.Wait()
	if waitErr != nil || extractErr != nil {
		_ = os.RemoveAll(dest)
		if waitErr != nil {
			return gitError("git archive failed", stderr.String(), waitErr)
		}
		return fmt.Errorf("tar extraction failed: %w", extractErr)
	}
	report(progress, "archive extracted to "+dest)
	return nil
}

// extractTar writes a tar stream into dest, refusing entries that would land
// outside it (guard: the archive comes from a remote, so its paths are
// untrusted).
func extractTar(r io.Reader, dest string) error {
	tr := tar.NewReader(r)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		name := filepath.Clean(filepath.FromSlash(hdr.Name))
		if name == "." {
			continue
		}
		target := filepath.Join(dest, name)
		if name == ".." || filepath.IsAbs(name) || !strings.HasPrefix(target, filepath.Clean(dest)+string(filepath.Separator)) {
			return fmt.Errorf("tar entry %q escapes the destination", hdr.Name)
		}
		switch hdr.Typeflag {
		case tar.TypeDir:
			if err := os.MkdirAll(target, 0o755); err != nil {
				return err
			}
		case tar.TypeReg:
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				return err
			}
			f, err := os.OpenFile(target, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
			if err != nil {
				return err
			}
			if _, err := io.Copy(f, tr); err != nil {
				_ = f.Close()
				return err
			}
			if err := f.Close(); err != nil {
				return err
			}
		default:
			// git archive emits directories and regular files for a tree.
		}
	}
}

// runGit runs one git command with argv (never a shell) under the git
// deadline and returns stdout/stderr as text.
func runGit(ctx context.Context, dir string, args ...string) (stdout, stderr string, err error) {
	ctx, cancel := gitContext(ctx)
	defer cancel()
	cmd := exec.CommandContext(ctx, "git", args...)
	if dir != "" {
		cmd.Dir = dir
	}
	var out, errBuf bytes.Buffer
	cmd.Stdout, cmd.Stderr = &out, &errBuf
	err = cmd.Run()
	return out.String(), errBuf.String(), err
}

// gitContext applies the per-invocation git deadline to a caller context.
func gitContext(ctx context.Context) (context.Context, context.CancelFunc) {
	if d := gitTimeout(); d > 0 {
		return context.WithTimeout(ctx, d)
	}
	return context.WithCancel(ctx)
}

// gitTimeout is the per-invocation git deadline: LEANKG_GIT_TIMEOUT seconds,
// or defaultGitTimeout. A non-positive value disables the deadline.
func gitTimeout() time.Duration {
	if v, ok := os.LookupEnv("LEANKG_GIT_TIMEOUT"); ok {
		if secs, err := strconv.Atoi(v); err == nil {
			return time.Duration(secs) * time.Second
		}
	}
	return defaultGitTimeout
}

// gitError renders the git diagnostic the way the Rust code did: the command
// context plus git's stderr, or the exec failure when git produced none.
func gitError(what, stderr string, err error) error {
	msg := strings.TrimSpace(stderr)
	if msg == "" {
		return fmt.Errorf("%s: %w", what, err)
	}
	return fmt.Errorf("%s: %s", what, msg)
}

// firstLine condenses a git diagnostic for a progress line.
func firstLine(stderr string, err error) string {
	msg := strings.TrimSpace(stderr)
	if msg == "" && err != nil {
		msg = err.Error()
	}
	if i := strings.IndexByte(msg, '\n'); i >= 0 {
		msg = msg[:i]
	}
	if msg == "" {
		msg = "no output"
	}
	return msg
}
